NFT Monitor Support

ijungmann opened the pull request:
**Description**:
Add NFT support to the mirror node monitor

* Add new fields to the TransactionSuppliers to allow for NFT transactions.
* Add support for nft expression that creates a token of type NON_FUNGIBLE

**Related issue(s)**:

Fixes #2132 

**Notes for reviewer**:
These have been tested against integration.

**Checklist**

- [x] Documented (Code comments, README, etc.)
- [x] Tested (unit, integration, etc.)

steven-sheehy commented on hedera-mirror-monitor/src/main/resources/application.yml:
```diff
@@ -1,14 +1,34 @@
+      network: OTHER
+      nodes:
+        - accountId: 0.0.3
+          host: 34.74.191.8
```
We shouldn't be exposing internal network IPs and private keys in temporary commits

ijungmann replied:
Weird, not sure how that got added to my commit, was not intending to push that file at all.  Removed.

ijungmann commented on docs/design/nft.md:
```diff
@@ -325,46 +325,26 @@ Optional Filters
 
 - Make changes to the `ExpressionConverter`
   - Add a new `nft` expression
```
Simplified this section now that we have a clearer picture.

ijungmann commented on docs/design/nft.md:
```diff
@@ -325,46 +325,26 @@ Optional Filters
 - Make changes to the `ExpressionConverter`
   - Add a new `nft` expression
-    - Creating an NFT requires the `tokenType` field to be set to `NON_FUNGIBLE_UNIQUE`. `TokenCreate` for an NFT also
-      cannot create serial numbers as part of the initial supply, so a `TokenMint` should follow to have serial numbers
-      created for the transaction suppliers.
```
Decided to remove the logic that made `nft` expressions mint a serial number as well, it made the ExpressionConverter messy and with the existing transfer/burn/wipe logic where we just continuously increment the serial number, one would already need to mint a huge amount of serial numbers to make those run at any scale, so this wasn't adding a lot of value.

ijungmann commented on docs/design/nft.md:
```diff
@@ -325,46 +325,26 @@ Optional Filters
-    the `amount` attribute or the `serial numbers` list.
-  - Because a serial number can only be transferred once out of a given account (unless it is transferred back), custom
-    logic will be needed for performant NFT transfers.
-    - One approach would be to have the supplier swap the sender and receiver each time to transfer the NFT back and
-      forth. This could be done only on the NFT transfer or on the entire transfer. This would likely limit performance
-      as the swap has to be synchronized (or we just let the transactions fail when double-transfers occur).
-      - This would also require different signatures on the transaction, which would require reworking how we handle the
-        transaction signing.
-    - Alernatively, the `ExpressionConverter` could just generate a large amount of serial numbers, and
-      the `CryptoTransferTransactionSupplier` could just transfer one at a time via a counter until it runs out. This
-      would be a much simpler approach, but it would cap the number of transactions possible in a scenario, and it would
-      require a longer startup time to generate all the serial nunmbers.
```
As mentioned in the sub-bullet here, swapping the accounts back and forth would have required having multiple signatures and swapping those out somewhere else.  As long as one runs a `MINT` scenario beforehand for a sufficient amount of time, the auto-increment seems to work well enough.

ijungmann commented on docs/design/nft.md:
```diff
@@ -325,46 +325,26 @@ Optional Filters
-      would be a much simpler approach, but it would cap the number of transactions possible in a scenario, and it would
-      require a longer startup time to generate all the serial nunmbers.
+  - Add a new `transferType` `NFT` that will allow for transferring NFTs. Change the name of `BOTH`
+    to `CRYTPO_AND_TOKEN`.
```
Almost certainly a better name for this.

ijungmann commented on hedera-mirror-datagenerator/src/main/java/com/hedera/datagenerator/sdk/supplier/account/CryptoTransferTransactionSupplier.java:
```diff
@@ -102,7 +112,15 @@ private void addTokenTransfers(TransferTransaction transferTransaction, TokenId
+
     public enum TransferType {
-        CRYPTO, TOKEN, BOTH
+        CRYPTO, TOKEN, NFT, CRYPTO_AND_TOKEN
```
I could also add a `CRYPTO_AND_NFT` (or something to that effect) if it seems useful.  Doing all 3 would require multiple token ids, could be done but would require a bit more configuring.

steven-sheehy replied:
My suggestion is only have CRYPTO, TOKEN and NFT in the enum then use a `Set<TransferType> transferTypes = new EnumSet<>()`.

ijungmann replied:
The set is a good idea.  The only thing about doing a TOKEN and an NFT transfer in one (on doing "ALL"), we would then require separate `nftTokenId` and `tokenId` (or something like that) so we know which goes in which list.  Otherwise it's pretty easy to add that option.

ijungmann replied:
I've added support for different combos via the `EnumSet`.  I had to add a `PropertiesCorrector` class that takes the list properties `transferTypes.0, transfersTypes.1` and puts them into a list, as they don't load correctly and thus don't get mapped to the supplier correctly.  Working on adding a test and some cleanup.

ijungmann commented on hedera-mirror-datagenerator/src/main/java/com/hedera/datagenerator/sdk/supplier/token/TokenBurnTransactionSupplier.java:
```diff
@@ -39,15 +42,31 @@
     private long maxTransactionFee = 1_000_000_000;
 
+    @NotNull
+    private final AtomicLong serialNumber = new AtomicLong(1); // The serial number to transfer.  Increments over time.
```
Comment may be excessive, but I wanted to note that the initial value is the starting serial number to burn/wipe/transfer, and it increments from there.

Nana-EC commented on hedera-mirror-datagenerator/src/main/java/com/hedera/datagenerator/sdk/supplier/account/CryptoTransferTransactionSupplier.java:
```diff
@@ -48,10 +50,13 @@
 
     @NotNull
-    private TransferType transferType = TransferType.CRYPTO;
+    private TransferType transferType = TransferType.NFT;
```
q: just curious of the choice NFT vs CRYPTO which doesn't require coordination of a token to first exist or an additional mint operation?

ijungmann replied:
That was leftover, just wanted to automatically have it do NFT transfers for a while and forgot to put it back.  Fixed now.

Nana-EC reviewed (changes requested):
Looks good, one nit and will do a final pass

steven-sheehy commented on hedera-mirror-monitor/src/main/resources/application.yml:
```diff
@@ -3,19 +3,18 @@ hedera:
       publish:
         scenarios:
-          pinger:
+          - name: HCS Pinger
```
Please undo changes to this file.

ijungmann replied:
Undone, apparently I reverted to a previous commit at some point.

steven-sheehy commented on hedera-mirror-monitor/src/main/java/com/hedera/mirror/monitor/expression/ExpressionConverterImpl.java:
```diff
@@ -84,6 +85,7 @@ private synchronized String doConvert(Expression expression) {
         try {
             log.debug("Processing expression {}", expression);
             ExpressionType type = expression.getType();
+            boolean isNft = type == ExpressionType.NFT;
```
Since there's no need for reuse, remove variable and move to if statement

ijungmann replied:
Good point, was leftover.  Fixed.

steven-sheehy commented on hedera-mirror-datagenerator/src/main/java/com/hedera/datagenerator/sdk/supplier/token/TokenWipeTransactionSupplier.java:
```diff
@@ -43,17 +46,33 @@
     @NotBlank
     private String tokenId;
 
+    TokenType type = TokenType.FUNGIBLE_COMMON;
```
private and not null

ijungmann replied:
Fixed.

steven-sheehy commented on hedera-mirror-datagenerator/src/main/java/com/hedera/datagenerator/sdk/supplier/token/TokenMintTransactionSupplier.java:
```diff
@@ -39,16 +43,34 @@
     @NotBlank
     private String tokenId;
 
+    TokenType type = TokenType.FUNGIBLE_COMMON;
```
private and not null

ijungmann replied:
Fixed.

steven-sheehy commented on hedera-mirror-datagenerator/src/main/java/com/hedera/datagenerator/sdk/supplier/token/TokenCreateTransactionSupplier.java:
```diff
@@ -48,28 +50,36 @@
     private int initialSupply = 1000000000;
 
+    @Min(1)
+    long maxSupply = 100000000000L;
```
private

ijungmann replied:
Fixed.

steven-sheehy commented on hedera-mirror-datagenerator/src/main/java/com/hedera/datagenerator/sdk/supplier/token/TokenCreateTransactionSupplier.java:
```diff
@@ -48,28 +50,36 @@
     @Min(1)
     private long maxTransactionFee = 1_000_000_000;
 
+    TokenSupplyType supplyType = TokenSupplyType.INFINITE;
```
private and not null this and `type`

ijungmann replied:
Fixed.

steven-sheehy commented on hedera-mirror-datagenerator/src/main/java/com/hedera/datagenerator/sdk/supplier/token/TokenBurnTransactionSupplier.java:
```diff
@@ -39,15 +42,31 @@
     @Min(1)
     private long maxTransactionFee = 1_000_000_000;
 
+    @NotNull
```
Not necessary on final fields. Same for other files.

ijungmann replied:
Fixed.  I'm now realizing a lot of the fields could be marked as final, I think in an earlier version of the monitor that was causing problems with loading the properties, but it seems to work now.

steven-sheehy commented on hedera-mirror-datagenerator/src/main/java/com/hedera/datagenerator/sdk/supplier/token/TokenBurnTransactionSupplier.java:
```diff
@@ -39,15 +42,31 @@
     @NotBlank
     private String tokenId;
 
+    TokenType type = TokenType.FUNGIBLE_COMMON;
```
private and not null

ijungmann replied:
Fixed.

steven-sheehy reviewed (changes requested):
Would be good to also update monitor.md with an example NFT flow since it seems like you need a mint operation at a sufficient tps for the other operations to also work. This dependency should be called out there.

xin-hedera reviewed (commented):
No issue found. Will sign off once the existing feedback is addressed.

steven-sheehy commented on hedera-mirror-datagenerator/src/main/java/com/hedera/datagenerator/sdk/supplier/account/CryptoTransferTransactionSupplier.java:
```diff
@@ -43,10 +45,12 @@
 
     @NotBlank
-    private String recipientAccountId;
+    private final String recipientAccountId;
```
This forces `recipientAccountId` and `senderAccountId` to be set via a constructor now, which I'm not sure will work with spring boot properties or be properly validated via the annotations that way. Let's just keep setters for everything that is settable by the user. Use final only for internal things that shouldn't be exposed.

ijungmann replied:
I did confirm that it is setting properly and validating when it is missing, so it can be done, but I'm fine with the standard you mentioned, especially since I didn't add `final` anywhere else so it's inconsistent.  I've removed it from these two fields.

ijungmann commented:
> Would be good to also update monitor.md with an example NFT flow since it seems like you need a mint operation at a sufficient tps for the other operations to also work. This dependency should be called out there.

Added a section for NFTs calling out the need to mint, also added a small blurb about List properties.

ijungmann commented on docs/design/nft.md:
```diff
@@ -325,46 +325,26 @@ Optional Filters
+  - `TokenMintTransactionSupplier` will need an enum `type` attribute, as well as a `metadata` string and
+    a `metadataSize` int (to generate metadata if not set, similar to how `ConsensusSubmitMessageTransactionSupplier`
+    works). The supplier should only set `amount` for fungible tokens, and only `metadata` for NFTs.
+  - `TokenBurnTransactionSupplier` and `TokenWipeTransactionSupplier` will need an enum `type` attribute, and it will
```
Removed.

Nana-EC commented on docs/monitor.md:
```diff
@@ -125,6 +145,32 @@ hedera:
+monitor requires a sufficient amount of serial numbers be minted prior to running any `CRYPTO_TRANSFER`
+, `TOKEN_BURN`, or `TOKEN_WIPE` scenario involving the NFT. A `TOKEN_MINT` scenario can be set up to accomplish this. If
+you are reusing an existing NFT, be sure to set the `serialNumber` property where applicable, as by default
+the `TransactionSupplier`will start with serial number 1, which may have already been deleted or transferred elsewhere.
```
nit: 
```suggestion
the `TransactionSupplier` will start with serial number 1, which may have already been deleted or transferred elsewhere.
```

ijungmann replied:
Fixed.

steven-sheehy commented on docs/monitor.md:
```diff
@@ -125,6 +145,32 @@ hedera:
+              tokenId: ${nft.1}
+              amount: 10
+            type: TOKEN_MINT
+
```
Add a nftTransfer to go with the mint and show tps for both. I'm interested in how the two TPS relate to each other. If I want to do 100 TPS nftTransfer how many TPS mints do I need?

You can probably move the `transferTypes` example in the previous section to here as the nftTransfer to combine both and simplify.

ijungmann replied:
Sure, I did some testing with both MINT and NFT_TRANSFER running at once.  To do 100 TPS on NFT_TRANSFER transferring one at a time, if you're also minting 1 NFT at a time, 105 TPS seems to be the stable minimum for MINT.  I added the TRANSFER to this example to show that, though it should be noted one could not run the example without some adjustments  (doesn't include TOKEN_ASSOCIATE, etc).

I also tested with transferring 1 NFT at 100 tps but minting 10 at a time, at around 12 TPS that looks stable after a few hiccups right at startup, so there are multiple ways to accomplish this.

steven-sheehy commented on hedera-mirror-datagenerator/src/main/java/com/hedera/datagenerator/sdk/supplier/account/CryptoTransferTransactionSupplier.java:
```diff
@@ -48,10 +55,14 @@
+
     @NotNull
-    private TransferType transferType = TransferType.CRYPTO;
+    private EnumSet<TransferType> transferTypes = EnumSet.of(CRYPTO);
```
```suggestion
    private Set<TransferType> transferTypes = EnumSet.of(CRYPTO);
```

ijungmann replied:
Fixed.

steven-sheehy commented on pom.xml:
```diff
@@ -65,6 +65,7 @@
         <grpc-spring-boot.version>2.11.0.RELEASE</grpc-spring-boot.version>
         <grpc.version>1.39.0</grpc.version>
         <guava.version>30.1.1-jre</guava.version>
+
```
nit: undo

ijungmann replied:
Forgot to push that, done.

ijungmann commented on hedera-mirror-monitor/src/main/java/com/hedera/mirror/monitor/properties/PropertiesCorrector.java:
```diff
@@ -0,0 +1,25 @@
+
+import java.util.Map;
+
+public interface PropertiesCorrector {
```
Suggestions for names are welcome.

steven-sheehy replied:
Remove it

Nana-EC replied:
Hmm, based on the logic in `PropertiesCorrectorImpl` it's more of an aggregator than it is a corrector right.
It's just generating a list value for list properties.

Maybe `ScenarioPropertiesAggregator` or `ScenarioPropertiesFormatter`.

ijungmann replied:
Yeah, it's a `corrector` in the sense that it comes in as a list but Jackson insists on serializing as individual properties, so this is "correcting" that, but I like `ScenarioPropertiesAggregator` as that's what it's really doing.  This will be removed at some point in a followup using the Deserializer Steven mentioned, but for now it will be `ScenarioPropertiesAggregator`

ijungmann commented on hedera-mirror-monitor/src/main/java/com/hedera/mirror/monitor/properties/PropertiesCorrectorImpl.java:
```diff
@@ -0,0 +1,59 @@
+                log.debug("Converting property {} into list {}", entry.getKey(), propertyName);
+                correctedProperties
+                        .merge(propertyName, Lists.newArrayList(entry.getValue()), (existingList, newList) -> {
+                            ((List) existingList).addAll((List) newList);
```
The casting here is a little gross, the alternative I had at one point was to have a separate Map of Lists and merge them into the `correctedProperties` map at the end, figured it was nicer to just have one map.

steven-sheehy commented on hedera-mirror-monitor/src/main/java/com/hedera/mirror/monitor/properties/PropertiesCorrectorImpl.java:
```diff
@@ -0,0 +1,59 @@
+            .compile("([A-Za-z0-9_]+)\\.[0-9]+");
+
+    @Override
+    public Map<String, Object> correctProperties(Map<String, String> properties) {
```
The correct way to attach custom deserialization logic to a field is to use `@JsonDeserialize(using = StringToSetDeserializer.class)`. This class becomes the `StringToSetDeserializer`. Then you're not leaking it into the generators and don't need to process the entire map, just a particular field.

ijungmann replied:
Maybe I'm missing something elsewhere, I commented out the `PropertiesCorrector`, added a simple `JsonDeserializer` and added it to the top of `transactionTypes`, and hit it with the debugger, but it never got into it, got the same issue:  `Unrecognized field "transferTypes.0"`.  It doesn't make the connection that the different elements belong to the same property, and instead create separate properties for each which then don't exactly match the property `transactionTypes`

steven-sheehy replied:
True. You'd have to change the `ScenarioProperties.properties` to a `Map<String, Object>`. Then it shows up as a `transferTypes` with a map as the field with "0" -> "CRYPTO". This would cause further refactoring of ExpressionConverter though. We can defer this as a follow up task.

ijungmann replied:
Got it, that makes sense.  Sure, I'll leave it and create a followup ticket for it.

Nana-EC commented on hedera-mirror-monitor/src/main/java/com/hedera/mirror/monitor/generator/ConfigurableTransactionGenerator.java:
```diff
@@ -99,8 +103,9 @@ public ConfigurableTransactionGenerator(ExpressionConverter expressionConverter,
 
     private TransactionSupplier<?> convert() {
         Map<String, String> convertedProperties = expressionConverter.convert(properties.getProperties());
+        Map<String, Object> correctedProperties = propertiesCorrector.correctProperties(convertedProperties);
```
Adding on to my aggregator/formatter naming comment below, I wonder if the `convert()` call should be moved into this new method call so it's a single aggregate/format call.

Not required as part of this PR. Just thinking out loud

ijungmann replied:
I thought about that, it would be a bit too much refactoring this late in the PR, and now from what Steven has shown we can remove that new class with some other refactoring, but I agree if this logic ever does get bigger it should move out of this class.

Nana-EC reviewed (commented):
Looks good, added naming suggestion. Will wait on other required signoff

Nana-EC reviewed (approved):
LGTM
